### PR TITLE
Refactor Classification Backbone to Factory Class Design

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

Dataset: CIFAR10-subset (custom)
Seed: 0
Deterministic: True
num_worker: 1
max_epochs: 3

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<div style='direction:ltr'>

Model | Before(releases/2.2.0) | After(This PR)
-- | -- | --
EfficientNet-B0 | train/loss:   0.428618     val/accuracy: 0.931001 | train/loss:   0.428618     val/accuracy: 0.931001
MobileNet-V3-Large | train/loss:0.426093   val/accuracy:   0.942757 | train/loss:0.426093   val/accuracy:   0.942757

</div>

</body>

</html>

### How to test

Integration Test
MULTI_CLASS_CLS
image
MULTI_LABEL_CLS
image
H_LABEL_CLS
image

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2024 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/classification/backbones/efficientnet.py [hunk 1]
```diff
@@ -439,17 +437,14 @@ def __init__(
...
         pooling_type: str | None = "avg",
         bn_eval: bool = False,
         bn_frozen: bool = False,
         instance_norm_first: bool = False,
-        pretrained: bool = False,
         **kwargs,
     ):
         super().__init__(**kwargs)
         self.num_classes = 1000
-        self.pretrained = pretrained
```
Comment (eunwoosh) [inline src/otx/algo/classification/backbones/efficientnet.py:452]: Could you remove docstring for this argument also?

**Review by eunwoosh — COMMENTED** [review]
LGTM. I have a minor comment. Please take a look :)

File: src/otx/algo/classification/backbones/efficientnet.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/classification/backbones/efficientnet.py:452]: Thanks - Done: https://github.com/openvinotoolkit/training_extensions/pull/3864/commits/3aafcefc3dfa9ab4f84533c5e4e824ec42a5e819

**Review by eunwoosh — APPROVED** [review]

File: src/otx/algo/classification/backbones/mobilenet_v3.py
Comment (sungchul1) [inline src/otx/algo/classification/backbones/mobilenet_v3.py:1]: Just a question, are `ModelInterface` and `MobileNetV3Base` necessary?
How about merging them into `MobileNetV3`?

File: src/otx/algo/classification/backbones/efficientnet.py [hunk 2]
```diff
@@ -528,164 +522,158 @@ def _init_params(self) -> None:
...
-        downsample = [1, 1, 1, 1, 0, 1, 0]
-        channels_per_layers = [16, 24, 40, 80, 112, 192, 320]
-        expansion_factors_per_layers = [1, 6, 6, 6, 6, 6, 6]
-        kernel_sizes_per_layers = [3, 3, 5, 3, 5, 5, 3]
-        _strides_per_stage = [1, 2, 2, 2, 1, 2, 1]
-        final_block_channels = 1280
-
-        layers = [int(math.ceil(li * depth_factor)) for li in layers]
-        channels_per_layers = [round_channels(ci * width_factor) for ci in channels_per_layers]
+    EFFICIENTNET_CFG: ClassVar[dict[str, Any]] = {
```
Comment (sungchul1) [inline src/otx/algo/classification/backbones/efficientnet.py:554]: Just a suggestion, how about using a common name such as `config` or `default_config` across all tasks?

**Review by sungchul1 — COMMENTED** [review]
Thanks for your hard work, I left minor comments.

File: src/otx/algo/classification/backbones/efficientnet.py (hunk 2 above)
Comment (harimkang) [reply src/otx/algo/classification/backbones/efficientnet.py:554]: actually, i follow litehrnet side : https://github.com/openvinotoolkit/training_extensions/blob/a5470fc3665228be8c3fde0bdc0728b11b26c057/src/otx/algo/segmentation/backbones/litehrnet.py#L1436
but we can discuss it :)

File: src/otx/algo/classification/backbones/mobilenet_v3.py
Comment (harimkang) [reply src/otx/algo/classification/backbones/mobilenet_v3.py:1]: Yeah i agree with you. I'd love to see those classes combined into one, but we need to carefully consider backward compatibility for now, so hopefully that can happen in the future step by step. (will do next PR)

**Review by sungchul1 — APPROVED** [review]

**Pull request merged**